File: SimulatorFrameworkBridgeLibTests/AccessibilityServiceServerTests.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <SimulatorFrameworkBridgeLib/AccessibilityService+Testing.h>

@interface AccessibilityServiceServerTests : XCTestCase
@end

@implementation AccessibilityServiceServerTests

- (void)testServeBacklogAccommodatesConcurrentProbes
{
  XCTAssertEqual(FBAXBridgeServeBacklogForTesting(), 16);
}

- (void)testExitOnDisconnectIsOffUnlessAsked
{
  XCTAssertFalse(FBAXBridgeExitOnDisconnectForTesting(@[]));
  XCTAssertFalse(FBAXBridgeExitOnDisconnectForTesting(@[@"--idle-timeout", @"30"]));
}

- (void)testExitOnDisconnectIsHonouredWhenAsked
{
  XCTAssertTrue(FBAXBridgeExitOnDisconnectForTesting(@[@"--exit-on-disconnect", @"1"]));
  XCTAssertTrue(FBAXBridgeExitOnDisconnectForTesting(@[@"--idle-timeout", @"30", @"--exit-on-disconnect", @"1"]));
}

- (void)testExitOnDisconnectCanBeTurnedOffExplicitly
{
  XCTAssertFalse(FBAXBridgeExitOnDisconnectForTesting(@[@"--exit-on-disconnect", @"0"]));
}

- (void)testAnIdleTimeoutOnTheCommandLineIsHonoured
{
  XCTAssertEqual(FBAXBridgeIdleTimeoutForTesting(@[@"--idle-timeout", @"45"], 300), 45);
}

- (void)testAServeWithNoIdleTimeoutUsesTheDefault
{
  XCTAssertEqual(FBAXBridgeIdleTimeoutForTesting(@[], 300), 300);
  XCTAssertEqual(FBAXBridgeIdleTimeoutForTesting(@[@"--something-else", @"1"], 300), 300);
  XCTAssertEqual(FBAXBridgeDefaultIdleTimeoutForTesting(), 300);
}

- (void)testAnUnusableIdleTimeoutFallsBackToTheDefault
{
  XCTAssertEqual(FBAXBridgeIdleTimeoutForTesting(@[@"--idle-timeout", @"0"], 300), 300);
  XCTAssertEqual(FBAXBridgeIdleTimeoutForTesting(@[@"--idle-timeout", @"-5"], 300), 300);
  XCTAssertEqual(FBAXBridgeIdleTimeoutForTesting(@[@"--idle-timeout", @"soon"], 300), 300);
  XCTAssertEqual(FBAXBridgeIdleTimeoutForTesting(@[@"--idle-timeout", @"12x"], 300), 300);
  XCTAssertEqual(FBAXBridgeIdleTimeoutForTesting(@[@"--idle-timeout"], 300), 300);
}

@end

File: SimulatorFrameworkBridgeLibTests/AccessibilityServiceServerTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
@_implementationOnly import SimulatorFrameworkBridgeLib
@_implementationOnly import SimulatorFrameworkBridgeSupport
import XCTest

final class AccessibilityServiceServerTests: XCTestCase {

  func testServeBacklogAccommodatesConcurrentProbes() {
    // A full backlog can reject a probe with ECONNREFUSED, which is indistinguishable from no listener.
    // Leave room for competing hosts to connect and learn that the single-threaded guest is busy.
    XCTAssertEqual(FBAXBridgeServer.serveBacklog, 16)
  }

  func testExitOnDisconnectIsOffUnlessAsked() {
    XCTAssertFalse(FBAXBridgeArguments.exitOnDisconnect(arguments: []))
    XCTAssertFalse(FBAXBridgeArguments.exitOnDisconnect(arguments: ["--idle-timeout", "30"]))
  }

  func testExitOnDisconnectIsHonouredWhenAsked() {
    XCTAssertTrue(FBAXBridgeArguments.exitOnDisconnect(arguments: ["--exit-on-disconnect", "1"]))
    XCTAssertTrue(FBAXBridgeArguments.exitOnDisconnect(arguments: ["--idle-timeout", "30", "--exit-on-disconnect", "1"]))
  }

  func testExitOnDisconnectCanBeTurnedOffExplicitly() {
    XCTAssertFalse(FBAXBridgeArguments.exitOnDisconnect(arguments: ["--exit-on-disconnect", "0"]))
  }

  func testAnIdleTimeoutOnTheCommandLineIsHonoured() {
    XCTAssertEqual(FBAXBridgeArguments.idleTimeout(arguments: ["--idle-timeout", "45"], fallback: 300), 45)
  }

  func testAServeWithNoIdleTimeoutUsesTheDefault() {
    XCTAssertEqual(FBAXBridgeArguments.idleTimeout(arguments: [], fallback: 300), 300)
    XCTAssertEqual(FBAXBridgeArguments.idleTimeout(arguments: ["--something-else", "1"], fallback: 300), 300)
    XCTAssertEqual(FBAXBridgeServer.defaultIdleTimeoutSeconds, 300)
  }

  func testAnUnusableIdleTimeoutFallsBackToTheDefault() {
    XCTAssertEqual(FBAXBridgeArguments.idleTimeout(arguments: ["--idle-timeout", "0"], fallback: 300), 300)
    XCTAssertEqual(FBAXBridgeArguments.idleTimeout(arguments: ["--idle-timeout", "-5"], fallback: 300), 300)
    XCTAssertEqual(FBAXBridgeArguments.idleTimeout(arguments: ["--idle-timeout", "soon"], fallback: 300), 300)
    XCTAssertEqual(FBAXBridgeArguments.idleTimeout(arguments: ["--idle-timeout", "12x"], fallback: 300), 300)
    XCTAssertEqual(FBAXBridgeArguments.idleTimeout(arguments: ["--idle-timeout"], fallback: 300), 300)
  }
}
